### PR TITLE
Implementing a full UI

### PR DESCRIPTION
**Pull request by tvwenger**
This is a work in progress. Just experimenting with React and UI components. Let me know what you think.

### CODE REVIEW

**Comment by tvwenger** [comment]
The latest commit fixes a logic error in `handleCatalysts`, makes ContainerNodes have the ability to require multiple containers to store the required `maintain` value, re-works the names a bit to reduce repetition in the visualization, and completes the implementation of both the step-by-step building instructions and full factory map (which is just a grid of step-by-step instructions). Let me know what you think!

File: package.json [hunk 1]
```diff
@@ -23,9 +23,11 @@
         "webpack-dev-server": "3.11.0"
     },
     "dependencies": {
+        "@types/react-svg-pan-zoom": "3.3.1",
```
Comment (lgfrbcsgo) [inline package.json:26]: The types should be a dev dependency.

File: src/graph.ts [hunk 2]
```diff
@@ -552,8 +630,7 @@ export class FactoryGraph {
      */
     createContainer(item: Item): ContainerNode {
         const containers = this.getContainers(item)
-        const name = item.name + " C" + containers.size
-        const container = new ContainerNode(name, item, this, 1.0)
+        const container = new ContainerNode("C" + containers.size, item, this, 1.0)
```
Comment (lgfrbcsgo) [inline src/graph.ts:633]: I would use a template literal here. Makes the absence of whitespaces immediately apparent.
```suggestion
        const container = new ContainerNode(`C${containers.size}`, item, this, 1.0)
```

File: src/graph.ts [hunk 3]
```diff
@@ -564,8 +641,7 @@ export class FactoryGraph {
      */
     createSplitContainer(item: Item, split: number): ContainerNode {
         const containers = this.getContainers(item)
-        const name = item.name + " C" + containers.size
-        const container = new ContainerNode(name, item, this, split)
+        const container = new ContainerNode("C" + containers.size, item, this, split)
```
Comment (lgfrbcsgo) [inline src/graph.ts:644]:
```suggestion
        const container = new ContainerNode(`C${containers.size}`, item, this, split)
```

File: src/graph.ts [hunk 4]
```diff
@@ -575,8 +651,11 @@ export class FactoryGraph {
      * @see {@link TransferContainerNode}
      */
     createTransferContainer(items: Item[]): TransferContainerNode {
-        const name = "Transfer Container " + this.transferContainers.size
-        const container = new TransferContainerNode(name, items, this)
+        const container = new TransferContainerNode(
+            "Trans Container" + this.transferContainers.size,
```
Comment (lgfrbcsgo) [inline src/graph.ts:655]:
```suggestion
            `Trans Container${this.transferContainers.size}`,
```

File: src/graph.ts [hunk 5]
```diff
@@ -587,8 +666,13 @@ export class FactoryGraph {
      */
     createOutput(item: Craftable, outputRate: PerMinute, maintainedOutput: Quantity): OutputNode {
         const containers = this.getContainers(item)
-        const name = item.name + " C" + containers.size
-        const output = new OutputNode(name, item, outputRate, maintainedOutput, this)
+        const output = new OutputNode(
+            "C" + containers.size,
```
Comment (lgfrbcsgo) [inline src/graph.ts:670]:
```suggestion
            `C${containers.size}`,
```

File: src/ui/new-factory.tsx [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+    const items = React.useMemo(() => values(ITEMS).filter(isCraftable), [ITEMS])
+    // NewFactory state and factory instructions
+    const [factoryState, setFactoryState] = React.useState<FactoryState>(FactoryState.SELECT)
+    const [factoryInstructions, setFactoryInstructions] = React.useState<FactoryInstruction[]>([])
+    // produced items, industry count, and maintain count
+    const [selection, setSelection] = React.useState<Craftable[]>([])
+    const [industryCount, setIndustryCount] = useMap<Craftable, number>()
+    const [maintainValue, setMaintainValue] = useMap<Craftable, number>()
+    // the FactoryGraph
+    const [factory, setFactory] = React.useState<FactoryGraph>()
```
Comment (lgfrbcsgo) [inline src/ui/new-factory.tsx:50]: We should probably refactor this into a single state object when we add another `useState`. This is fine for now.

**Review by lgfrbcsgo — APPROVED** [review]
Wow. This is awesome!

Some of the rendering should probably be broken up into smaller components, but I would just merge this first and clean up later.

Maybe I can deploy this later.

File: package.json (hunk 1 above)
Comment (tvwenger) [reply package.json:26]: I tried to install this as a dev dependency, and I couldn't get it to compile. Maybe you will have more luck.

File: src/graph.ts (hunk 2 above)
Comment (tvwenger) [reply src/graph.ts]: Fixed in https://github.com/lgfrbcsgo/du-factory-generator/commit/e3612727cf42e8f92e923180d0c72033bd0a1513

File: src/graph.ts (hunk 3 above)
Comment (tvwenger) [reply src/graph.ts]: Fixed in https://github.com/lgfrbcsgo/du-factory-generator/commit/e3612727cf42e8f92e923180d0c72033bd0a1513

File: src/graph.ts (hunk 4 above)
Comment (tvwenger) [reply src/graph.ts]: Fixed in https://github.com/lgfrbcsgo/du-factory-generator/commit/e3612727cf42e8f92e923180d0c72033bd0a1513

File: src/graph.ts (hunk 5 above)
Comment (tvwenger) [reply src/graph.ts]: Fixed in https://github.com/lgfrbcsgo/du-factory-generator/commit/e3612727cf42e8f92e923180d0c72033bd0a1513

**Comment by tvwenger** [comment]
Added missing ore containers

**Pull request merged**